### PR TITLE
WT-6575 Call Tag Upsert API when adding keywords for a case

### PR DESCRIPTION
**Pull request by alvinllobrera**

### CODE REVIEW

**Comment by vyshane** [comment]
@sukolenvo, @Galya-IT, this was made necessary because we want to upsert a tag if it doesn't yet exist and we are trying to add keywords to it. See WT-6575 for context.

Tag sync and keyword sync are asynchronous. Consider:

- We drain all new cases, creating a tag for each
- We start creating keywords, but there are a lot of them
- In the mean time, a new case is added to the db, but we haven't finished with keyword creation
- A FileSite project is created for the case too, and we detect that one
- We attempt to create the new keyword, during the same batch keyword run, while we haven't created its case yet

Trying to create the keyword will fail since WiseTime doesn't yet know about its case. However, if we add the tag at the same time as creating the keyword, then all is good.

File: src/test/java/io/wisetime/connector/inprotech/InprotechConnectorPerformTagUpdateTest.java [hunk 1]
```diff
@@ -345,18 +361,23 @@ void performTagUpdate_new_projects_found() throws IOException {
...
-        .as("We should update tag keywords with new project ids")
-        .containsExactlyInAnyOrder(
-            ImmutableSet.of("FID" + project3.getId()),
-            ImmutableSet.of("FID" + project1.getId(), "FID" + project2.getId()));
-    assertThat(addKeywordsRequests.getValue().keySet())
-        .containsExactlyInAnyOrder(irn2, irn1);
+    ArgumentCaptor<List<UpsertTagRequest>> upsertTagRequestArgCaptor = ArgumentCaptor.forClass(List.class);
+    verify(apiClientMock, times(1)).tagUpsertBatch(upsertTagRequestArgCaptor.capture());
+
+    assertThat(upsertTagRequestArgCaptor.getValue().size())
```
Comment (sukolenvo) [inline src/test/java/io/wisetime/connector/inprotech/InprotechConnectorPerformTagUpdateTest.java]:
```java
assertThat(upsertTagRequestArgCaptor.getValue())
  .as("Only 1 case is valid from the detected keywords")
  .hasSize(1);
```
Will give us more detailed report with exception

**Comment by Galya-IT** [comment]
> Trying to create the keyword will fail since WiseTime doesn't yet know about its case. However, if we add the tag at the same time as creating the keyword, then all is good.

I'm not completely sure how the `upsert` works. 
A case gets added as a tag because of its corresponding FileSite project being detected prior to the case itself. Case IRN and FileSite project Id are set as keywords.
When `performNewTagDetection` comes to the same case that's already added and the upsert request sends only name and description of the case and implicitly the name as additionalKeyword, are the keywords going to be overwritten?

**Comment by vyshane** [comment]
> > Trying to create the keyword will fail since WiseTime doesn't yet know about its case. However, if we add the tag at the same time as creating the keyword, then all is good.
> 
> I'm not completely sure how the `upsert` works.
> A case gets added as a tag because of its corresponding FileSite project being detected prior to the case itself. Case IRN and FileSite project Id are set as keywords.
> When `performNewTagDetection` comes to the same case that's already added and the upsert request sends only name and description of the case and implicitly the name as additionalKeyword, are the keywords going to be overwritten?

The existing keywords will not be overwritten. The upsert request will add them as additional keywords.

File: src/main/java/io/wisetime/connector/inprotech/InprotechDao.java [hunk 2]
```diff
@@ -85,6 +85,15 @@
         .listResult(this::buildCaseFromResultSet);
   }
 
+  List<Case> getValidCasesByIrn(Set<String> caseIrns) {
```
Comment (vyshane) [inline src/main/java/io/wisetime/connector/inprotech/InprotechDao.java]: How about `getActiveCasesByIrn`?

**Pull request merged**